getnacelle/nacelle-js
# feat(gatsby)!: infer types for `Content.fields`

## Why are these changes introduced?

### Ticket

[ENG-4001](https://nacelle.atlassian.net/browse/ENG-4001)

### Problem statement

[`@nacelle/gatsby-source-nacelle`](https://www.gatsbyjs.com/plugins/@nacelle/gatsby-source-nacelle) version `8.x` could not source `remoteImage` data from Nacelle v2 `Content` due to image fields being held in sub-properties of `Content.fields`, which is presently typed as `JSON` in the Nacelle Storefront GraphQL schema. Because those fields were typed as `JSON`, Gatsby didn't have enough knowledge of the `Content` data to be able to create [image nodes](https://www.gatsbyjs.com/docs/how-to/images-and-media/preprocessing-external-images) for [`gatsby-plugin-image`](https://www.gatsbyjs.com/plugins/gatsby-plugin-image) at arbitrary data depths.

## What is this pull request doing?

### Main thrust

This PR is the culmination of #104 and #121. By leveraging [Gatsby schema inference](https://www.gatsbyjs.com/docs/schema-inference), we add rich type information to content data. Some benefits and implications of this:

#### Images

Gatsby image nodes can now be sourced at arbitrary depths in content queries.

#### Querying content data

Rich type information enables merchant developers to make granular queries for the fields of interest. This is likely to offer a significant reduction in payload sizes for content queries. Implications include:
1. Instead of one `NacelleContent` type, types are now created for each content model. These content types are bespoke to a Nacelle space.
2. Merchant developers have access to new root-level queries corresponding to their bespoke content types.
3. When querying arrays of content data that contain multiple content types, [inline fragments](https://graphql.org/learn/queries/#inline-fragments) must be used to select fields for each type.

### Supporting & miscellaneous changes

1. In the `README`s for `starters/gatsby` and `packages/gatsby-source-nacelle`, add links to Gatsby docs which explain the new query patterns, including inline fragments. The docs updates will be published after this PR is merged, but they can be previewed [here](https://nacelle.atlassian.net/wiki/spaces/PT/pages/1613627437/Guide+Using+Gatsby#Data-Fetching).
4. Updates `packages/gatsby-source-nacelle/LICENSE` to use the `nacelle-js` Apache-2.0 license.

## 

### How to Test

Many of the instructions below are borrowed from #104.

1. Check out the `ENG-4001/gatsby-source-nacelle-9.0` branch.
4. `npm run bootstrap` in the monorepo root if you haven't already.
5. `cd starters/gatsby`
6. Set `.env` with the `GATSBY_NACELLE_STOREFRONT_ENDPOINT` and `GATSBY_NACELLE_STOREFRONT_TOKEN` from a Nacelle v2 space of your choosing.
7. `npm run develop`
8. Open the [GraphiQL](https://www.gatsbyjs.com/docs/how-to/querying-data/running-queries-with-graphiql) endpoint at [`localhost:8000/___graphql`](http://localhost:8000/___graphql). 
9. For CMS content, you should see query types for each of your CMS content types prefixed with "`NacelleContentRemote`" and be able to select individual fields from the `remoteFields` key, as well as have RemoteImage files for each image.
10. For other content types, you will need to query the top level type (e.g. `Product`, `ProductCollection`, `ContentCollection`, etc.) and then under the `content` field, be able to query for individual `remoteFields`, including `remoteImage` fields for any image nodes.

Queries for content data, including content data containing images, should work as expected at any query depth.

### Testing checklist

* [x] Updated public facing documentation (pending)
